core/modules/shrscan/Director.h

```diff
@@ -0,0 +1,315 @@
+//! @param  lkData    whether or not to lock the table's data
+//-----------------------------------------------------------------------------
+
+            TableInfo(const char *tabName="", bool lkIndex=true, bool lkData=true)
+                     : tableName(tabName), lockIndex(lkIndex), lockData(lkData)
+                     {}
+
+//-----------------------------------------------------------------------------
+//! Destructor
+//-----------------------------------------------------------------------------
+
+           ~TableInfo() {}
+};
+
+//-----------------------------------------------------------------------------
+//! @brief The Query pure abstract class describes a shared scan query.
+//-----------------------------------------------------------------------------
+
+class Query
+{
+public:
+
+//-----------------------------------------------------------------------------
+//! @brief The callback method to indicate a query has finished its run.
+//!
+//! The endQuery() method is called when the query completes or is canceled.
+//! This method is always called on a separate thread.
+//!
+//! @param  badChunks is a smart pointer to a vector of unprocessed chunks. This
+//!                   includes the chunk that was being processed if Run()
+//!                   returned false. You must take ownership of the shared
+//!                   pointer if you want to persist the vector upon return.
+//! @return canceled  When true, the query was cancelled and badChunks will be a
+//!                   zero length vector. Otherwise, the query completed and
+//!                   badChunks does hold the unprocessed chunk numbers.
+//-----------------------------------------------------------------------------
+
+virtual void endQuery(std::shared_ptr<std::vector<int>> const& badChunks,
+                      bool canceled) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief The callack method to run a query on a chunk.
+//!
+//! The runQuery() method is invoked by the shared scan director to initiate the
+//! start of a query on a particular chunk.
+//!
+//! @param  chunkNum  is the number of the chunk to be processed.
+//!
+//! @return false     the query should be cancelled and endQuery() called.
+//! @return true      continue to the next chunk, if any.
+//-----------------------------------------------------------------------------
+
+virtual bool runQuery(int chunkNum) = 0;
+
+//-----------------------------------------------------------------------------
+//! The following vector should be used to add additional tables that may be
+//! needed by this particular query prior to call Director::addQuery(). These
+//! are in addition to the tables that the director was created with.
+//-----------------------------------------------------------------------------
+
+std::vector<TableInfo> tables; //< Additional tables needed by this query
```

Comment:
This is some mix of abstract interface ans concrete data. Can we keep everything abstract without assumptions that it needs to have a vector inside it?